data-processing/entities/abbreviations/extractor.py

```diff
@@ -0,0 +1,163 @@
+# ⎋: https://github.com/nipunsadvilkar/pySBD/blob/master/pysbd/punctuation_replacer.py
+# ᓰ: https://github.com/nipunsadvilkar/pySBD/blob/master/pysbd/punctuation_replacer.py
+# ᓱ: https://github.com/nipunsadvilkar/pySBD/blob/master/pysbd/punctuation_replacer.py
+# ᓳ: https://github.com/nipunsadvilkar/pySBD/blob/master/pysbd/punctuation_replacer.py
+# ᓴ: https://github.com/nipunsadvilkar/pySBD/blob/master/pysbd/processor.py, https://github.com/nipunsadvilkar/pySBD/blob/master/pysbd/punctuation_replacer.py
+# ᓷ: https://github.com/nipunsadvilkar/pySBD/blob/master/pysbd/cleaner.py, https://github.com/nipunsadvilkar/pySBD/blob/master/pysbd/punctuation_replacer.py
+# ᓸ: https://github.com/nipunsadvilkar/pySBD/blob/master/pysbd/processor.py, https://github.com/nipunsadvilkar/pySBD/blob/master/pysbd/punctuation_replacer.py
+PYSBD_RESERVED_CHARACTERS: List[str] = [
+    "∯",
+    "ȸ",
+    "♨",
+    "☝",
+    "✂",
+    "⎋",
+    "ᓰ",
+    "ᓱ",
+    "ᓳ",
+    "ᓴ",
+    "ᓷ",
+    "ᓸ",
+]
+
+symbols = ["%", "^", "{", "}", "[", "]", "\\", "=", "#", "&", "~", "$", "|", "_", ":", ";"]
+
+class AbbreviationExtractor(EntityExtractor):
+    """
+    Extract plaintext sentences from TeX, with offsets of the characters they correspond to in
+    the input TeX strings. The extracted sentences might include some junk TeX, having the same
+    limitations as the plaintext produced by PlaintextExtractor.
+    """
+
+    def parse(self, tex_path: str, tex: str) -> Iterator[Abbreviation]:
+        for reserved_char in PYSBD_RESERVED_CHARACTERS:
+            if reserved_char in tex:
+                logging.warning(
+                    'Reserved character from pysbd "%s" found in tex string, this might break the sentence extractor.',
+                    reserved_char,
+                )
+
+        # Extract plaintext segments from TeX
+        plaintext_extractor = PlaintextExtractor()
+        plaintext_segments = plaintext_extractor.parse(tex_path, tex)
+
+        # Build a map from character offsets in the plaintext to TeX offsets. This will let us
+        # map from the character offsets of the sentences returned from the sentence boundary
+        # detector back to positions in the original TeX.
+        plaintext_to_tex_offset_map = {}
+        plaintext = ""
+        last_segment = None
+        for segment in plaintext_segments:
+            for i in range(len(segment.text)):
+                tex_offset = (
+                    (segment.tex_start + i)
+                    if not segment.transformed
+                    else segment.tex_start
+                )
+                plaintext_to_tex_offset_map[len(plaintext) + i] = tex_offset
+
+            # While building the map, also create a contiguous plaintext string
+            plaintext += segment.text
+            last_segment = segment
+
+        if last_segment is not None:
+            plaintext_to_tex_offset_map[len(plaintext)] = last_segment.tex_end
```

Comment:
If I recall correctly, this code also appears in the sentence extractor. Would it make sense to factor it out into another file (say, `common/parse_tex.py`) so that the code can be maintained in just one place?

Reply:
Yes, it is. I guess it might be better to factor it out.